### PR TITLE
CORTX-28892 : Write wrappers to get disk and cvg ids from gconf

### PR DESCRIPTION
**Pull request by akash2144**
Signed-off-by: Akash Dudhane <akash.a.dudhane@seagate.com>

# Problem Statement
- Write wrappers to get disk and cvg ids from gconf

# Design
-  Added 3 functions to:
Return list of CVGs for any given node using search APIs
Return list of disks for any given node and CVG using search APIs
Return list of disks for any given node using search APIs

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: N
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N]: NA
- [ ] Side effects on other features (deployment/upgrade)? [Y/N] N
- [ ] Dependencies on other component(s)? [Y/N] N
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by akash2144** [comment]
**Test Details :** 

By calling API from HA utils we will get the list of CVG's & Disks
```
[root@cortx-ha-headless-svc /]# python3 test_conf_store_search_api.py
2022-03-03 12:06:48 test_conf_store_search_api [89]: INFO [init] Started logging for service test_conf_store_search_api
******** Testing confstore search APIs for data and server PODs ********
2022-03-03 12:06:48 test_conf_store_search_api [89]: INFO [set_cluster_cardinality] Cluster cardinality: number of nodes 6, machine ids for nodes ['a94b5143999447aeb1f735b6e53931c4', 'f32db33f32a54927a91e52e3ec9c491a', '4368536608504e3a973dd16c066884ff', '39c238927338485c8bccfcbe1effe68f', '73aadfc83aab4cd1ad965ac2825345af', '21d6291109304485b3daff43a06cff77']
Following clsuter cardinality is set: (6, ['a94b5143999447aeb1f735b6e53931c4', 'f32db33f32a54927a91e52e3ec9c491a', '4368536608504e3a973dd16c066884ff', '39c238927338485c8bccfcbe1effe68f', '73aadfc83aab4cd1ad965ac2825345af', '21d6291109304485b3daff43a06cff77'])


**************   Testing confstore search APIs for CVG & disk  ***********


CVG's presents under node 21d6291109304485b3daff43a06cff77 are :  ['cvg-01', 'cvg-02']

Disks present under node 21d6291109304485b3daff43a06cff77 are :  ['/dev/sdd', '/dev/sde', '/dev/sdc', '/dev/sdg', '/dev/sdh', '/dev/sdf']

Disk present under node 21d6291109304485b3daff43a06cff77 & CVG cvg-01 are :  ['/dev/sdd', '/dev/sde', '/dev/sdc']
```

**Review by Madhura-08 — COMMENTED** [review]
Please fix codacy issues

**Review by Madhura-08 — APPROVED** [review]
As discussed, Approving it. Please check with Archana regarding those two minor codacy issues

File: ha/test/integration/test_conf_store_search_api.py [hunk 1]
```diff
@@ -24,14 +24,27 @@
...
+        print("******** Testing confstore search APIs for data and server PODs ********")
         cluster_card.set_cluster_cardinality("cortx")
         cc = cluster_card.get_cluster_cardinality()
         print(f"Following clsuter cardinality is set: {cc}")
 
+        print("\n\n**************   Testing confstore APIs for CVG & disk  ***********\n")
+        node_id = cluster_card._get_data_pods("cortx")[0]
+
+        cvg_list = cluster_card.get_cvg_list("cortx", node_id)
+        print(f"\nCVG's presents under node {node_id} are : ", cvg_list)
```
Comment (indrajitzagade) [inline ha/test/integration/test_conf_store_search_api.py:41]: *CVGs present

Comment (akash2144) [reply ha/test/integration/test_conf_store_search_api.py]: Done.

File: ha/util/conf_store.py [hunk 2]
```diff
@@ -110,3 +121,93 @@ def set_cluster_cardinality(self, index):
...
+                                            replace("(node_id)", node_id).replace("(cvg_index)", str(cvg_index)))
+                num_of_metadata_disks = Conf.get(index, GconfKeys.METADATA_COUNT.value.
+                                                replace("(node_id)", node_id).replace("(cvg_index)", str(cvg_index)))
+
+                for d_index in range(num_of_data_disks):
+                    disk_list.append(Conf.get(index, GconfKeys.DATA_DISK.value.replace("(node_id)", node_id).
+                                              replace("(cvg_index)", str(cvg_index)).replace("(d_index)", str(d_index))))
+
+                for m_index in range(num_of_metadata_disks):
+                    disk_list.append(Conf.get(index, GconfKeys.METADATA_DISK.value.replace("(node_id)", node_id).
```
Comment (mariyappanp) [inline ha/util/conf_store.py:208]: [Suggestion]
As pep8 recommend, limit the line length to 79 characters.

**Review by mukhtar-inamdar — APPROVED** [review]

File: ha/util/conf_store.py (hunk 2 above)
Comment (akash2144) [reply ha/util/conf_store.py:208]: Yes, You're right. I already reduced the length of lines wherever possible. If still we reduce the length of lines then it will unnecessarily increase the no. of lines.

File: ha/util/conf_store.py [hunk 3]
```diff
@@ -29,6 +30,17 @@
...
 
+
+# Confstore keys
+class GconfKeys(enum.Enum):
+    CVG_NAME = "node>(node_id)>storage>cvg[(cvg_index)]>name"
+    CVG_COUNT = "node>(node_id)>storage>num_cvg"
+    DATA_COUNT = "node>(node_id)>storage>cvg[(cvg_index)]>devices>num_data"
+    METADATA_COUNT = "node>(node_id)>storage>cvg[(cvg_index)]>devices>num_metadata"
+    DATA_DISK = "node>(node_id)>storage>cvg[(cvg_index)]>devices>data[(d_index)]"
+    METADATA_DISK = "node>(node_id)>storage>cvg[(cvg_index)]>devices>metadata[(m_index)]"
```
Comment (mariyappanp) [inline ha/util/conf_store.py:41]: Instead of replacing the str, we can use placeholder. This will avoid accidently replacing matching string.
Ex:
```
CVG_NAME = f"node>{node_id}>storage>cvg[{cvg_index}]>name"

CVG_NAME.format(node_id=1, cvg_index=2)
```

We need not to even convert any data to str.

File: ha/util/conf_store.py [hunk 4]
```diff
@@ -110,3 +121,93 @@ def set_cluster_cardinality(self, index):
...
+        """
+        try:
+            cvg_list = []
+            cvg_count = Conf.get(index, GconfKeys.CVG_COUNT.value.replace("(node_id)", node_id))
+            cvg_list = [Conf.get(index, GconfKeys.CVG_NAME.value.replace("(node_id)", node_id)
+                            .replace("(cvg_index)", str(c))) for c in range(cvg_count)]
+            return cvg_list
+        except Exception as e:
+            Log.error(f"Unable to fetch CVG list from GConf. Error {e}")
+            raise Exception(f"Unable to fetch CVG list. Error {e}")
```
Comment (mariyappanp) [inline ha/util/conf_store.py:145]: Conf will return `cvg_count` as None for non-existing node_id. Then we will see `TypeError: None type object` from line 141. At last, the error message won't be relevant to node_id.
If cvg_count is None from Conf, raise Exception with appropriate message.

File: ha/util/conf_store.py [hunk 5]
```diff
@@ -110,3 +121,93 @@ def set_cluster_cardinality(self, index):
...
+            index(str): index of conf file
+            node_id (str): node id
+        Returns:
+            list: list of disks id's
+
+        >>> get_disk_list(index, '21d6291109304485b3daff43a06cff77')
+        ['/dev/sdd', '/dev/sde', '/dev/sdc', '/dev/sdg', '/dev/sdh', '/dev/sdf']
+        """
+        try:
+            cvg_count = Conf.get(index, GconfKeys.CVG_COUNT.value.replace("(node_id)", node_id))
```
Comment (mariyappanp) [inline ha/util/conf_store.py:160]: Same as my previous comment. Ref: https://github.com/Seagate/cortx-ha/pull/660/files#r819503014

File: ha/util/conf_store.py (hunk 2 above)
Comment (mariyappanp) [reply ha/util/conf_store.py:208]: Fine.

File: ha/util/conf_store.py (hunk 3 above)
Comment (akash2144) [reply ha/util/conf_store.py:41]: No, we can't use formatting here.
Without specifying node_id & cvg_index we can't use formatting here _f"node>{node_id}>storage>cvg[{cvg_index}]>name"_

File: ha/util/conf_store.py [hunk 6]
```diff
@@ -110,3 +121,93 @@ def set_cluster_cardinality(self, index):
         cluster_cardinality_key = CLUSTER_CARDINALITY_KEY
         cluster_cardinality_value = {CLUSTER_CARDINALITY_NUM_NODES: num_pods, CLUSTER_CARDINALITY_LIST_NODES : watch_pods }
         self._confstore.update(cluster_cardinality_key, json.dumps(cluster_cardinality_value))
+
+    def get_cvg_list(self, index, node_id):
```
Comment (mariyappanp) [inline ha/util/conf_store.py:125]: Method can be static_method. It would be appropriate as it is utils.

File: ha/util/conf_store.py (hunk 3 above)
Comment (mariyappanp) [reply ha/util/conf_store.py:41]: My bad, avoid f"". But still that will work, @akash2144 
```
>>>
>>> CVG_NAME = "node>{node_id}>storage>cvg[{cvg_index}]>name"
>>> CVG_NAME.format(node_id=1, cvg_index=2)
'node>1>storage>cvg[2]>name'
>>>
```

Comment (akash2144) [reply ha/util/conf_store.py]: Done, Thanks!

File: ha/util/conf_store.py (hunk 4 above)
Comment (akash2144) [reply ha/util/conf_store.py]: Raised Exception with appropriate message.

File: ha/util/conf_store.py (hunk 5 above)
Comment (akash2144) [reply ha/util/conf_store.py]: Raised Exception with appropriate message.

File: ha/util/conf_store.py (hunk 6 above)
Comment (akash2144) [reply ha/util/conf_store.py]: Done.

File: ha/util/conf_store.py [hunk 7]
```diff
@@ -29,6 +30,17 @@
 from cortx.utils.log import Log
 from ha.core.error import ClusterCardinalityError
 
+
+# Confstore keys
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:34]: Some keys are already defined in k8s_setup/const.py 
Can you please have all keys in the same place

File: ha/util/conf_store.py [hunk 8]
```diff
@@ -29,6 +30,17 @@
...
+class GconfKeys(enum.Enum):
+    CVG_NAME = "node>{node_id}>storage>cvg[{cvg_index}]>name"
+    CVG_COUNT = "node>{node_id}>storage>num_cvg"
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:37]: Use the DELIM const for '>'

File: ha/util/conf_store.py [hunk 9]
```diff
@@ -110,3 +121,96 @@ def set_cluster_cardinality(self, index):
         cluster_cardinality_key = CLUSTER_CARDINALITY_KEY
         cluster_cardinality_value = {CLUSTER_CARDINALITY_NUM_NODES: num_pods, CLUSTER_CARDINALITY_LIST_NODES : watch_pods }
         self._confstore.update(cluster_cardinality_key, json.dumps(cluster_cardinality_value))
+
+    @staticmethod
+    def get_cvg_list(index, node_id):
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:126]: There are many functionalities common across these 3 functions.
See if you can create a separate functions for the common functionalities  and use them in these 3 functions

File: ha/util/conf_store.py [hunk 10]
```diff
@@ -110,3 +121,96 @@ def set_cluster_cardinality(self, index):
...
+            list: list of CVG's
+
+        >>> ConftStoreSearch.get_cvg_list("cortx", '21d6291109304485b3daff43a06cff77')
+        ['cvg-01', 'cvg-02']
+        """
+        cvg_list = []
+        try:
+            cvg_count = Conf.get(index, GconfKeys.CVG_COUNT.value.format(node_id=node_id))
+            if cvg_count is None:
+                raise Exception(f"CVGs are not available for this node {node_id}")
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:142]: I don't think exception is needed here.  We can log an error message and continue.
@indrajitzagade  @shaileshsvc  any comments ?

File: ha/util/conf_store.py [hunk 11]
```diff
@@ -110,3 +121,96 @@ def set_cluster_cardinality(self, index):
...
+            cvg_id (str): cvg id
+        Returns:
+            list: list of disks id's
+
+        >>> ConftStoreSearch.get_disk_list_for_cvg("cortx", '21d6291109304485b3daff43a06cff77', 'cvg-01')
+        ['/dev/sdd', '/dev/sde', '/dev/sdc']
+        """
+        disk_list = []
+        try:
+            cvg_list = ConftStoreSearch.get_cvg_list(index, node_id)
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:200]: Why is this needed ? cvg_id is already provided as input to the function

File: ha/util/conf_store.py [hunk 12]
```diff
@@ -110,3 +121,96 @@ def set_cluster_cardinality(self, index):
...
+            if cvg_id in cvg_list:
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:201]: We can use the search APIs to get the full keys for the given CVG and values for the associated data/metadata disks right ?

**Review by mariyappanp — APPROVED** [review]

File: ha/util/conf_store.py (hunk 10 above)
Comment (mariyappanp) [reply ha/util/conf_store.py:142]: IMO, returning empty list would be fine. Let the client raise it on need.

File: ha/util/conf_store.py (hunk 12 above)
Comment (ArchanaLimaye) [reply ha/util/conf_store.py:201]: Discussed with akash. If we are to use search APIs, two iterations will have to be done. 
1. To get keys of all cvgs matching the cvg_id; across all nodes. ( there is no way to specify a particular node while searching for cvg_ids in search API)
2. To find out the correct key from 1 
3. To get disks associated with this CVG
so no benefit of using search APIs here.
hence closing the comment

File: ha/util/conf_store.py (hunk 7 above)
Comment (akash2144) [reply ha/util/conf_store.py]: Done. Moved Confstore class to k8s_setup/const.py

File: ha/util/conf_store.py (hunk 8 above)
Comment (akash2144) [reply ha/util/conf_store.py]: Done.

File: ha/util/conf_store.py (hunk 9 above)
Comment (akash2144) [reply ha/util/conf_store.py:123]: Done.

File: ha/util/conf_store.py (hunk 10 above)
Comment (akash2144) [reply ha/util/conf_store.py]: Logging exception message with Log.warn & returning empty list in this case.

File: ha/util/conf_store.py (hunk 11 above)
Comment (akash2144) [reply ha/util/conf_store.py:211]: As discussed, To find out cvg_index for cvg_id.

File: ha/util/conf_store.py [hunk 13]
```diff
@@ -110,3 +110,111 @@ def set_cluster_cardinality(self, index):
...
+        cvg_list = []
+        try:
+            cvg_count = ConftStoreSearch._get_cvg_count(index, node_id)
+            if cvg_count:
+                cvg_list = [Conf.get(index, GconfKeys.CVG_NAME.value.format
+                                     (_DELIM=_DELIM, node_id=node_id, cvg_index=cvg_index)) for cvg_index in range(cvg_count)]
+            return cvg_list
+        except Exception as e:
+            Log.error(f"Unable to fetch CVG list from GConf. Error {e}")
+            raise Exception(f"Unable to fetch CVG list. Error {e}")
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:143]: Please confirm if raising exception is fine or this needs to be a error log.
Same comment for all functions

**Review by ArchanaLimaye — APPROVED** [review]
